### ISSUE TITLE
RFC: Cloud Support for Scheduler

### ISSUE THREAD

**Issue by mitchelsellers** [dnnsoftware/Dnn.Platform#2673]
## Description of Problem
Currently, the DNN Scheduler system runs in-process within a DNN installation and has a configuration option for "Run on Server(s)".  In cloud environments, such as Azure App Service, with auto-scale situations, it is not possible to configure a scheduled job to only execute on a single server.

This is an issue when jobs interact with the database or otherwise, that should only be working with a single instance.  (DB Maintenance, email distribution, etc.) 

A simple method for "Run once" or similar is needed

## Proposed Solution Option 1
The existing Scheduler Interface should be updated to have a new optional setting for "Run Once Per Schedule".  Or a similar name.  This setting should then prevent the application from running the job more than once at the same time.

Implementation of this solution is a bit unclear and would need research.  However, this solution would/should be platform agnostic

## Proposed Solution Option 2 
Another solution would be to implement an environment specific (Azure or otherwise) solution that leverages technologies that exist in the environment.

This isn't preferred as it would be a single solution for an environment.

## Affected version
All Version

## Requested Feedback
Does anyone have other ideas of options, or methods that we could utilize?

**Comment by ahoefling** [comment]
While I have not looked at the Scheduled Job code in DNN I think we should consider creating an extension point for the scheduler to use Azure WebJobs. Perhaps implement it like JWT as an optional feature that a host user can turn on/off. This way if someone if hosting in Azure they can automatically use the power of Azure WebJobs which will handle this use case. 

The downside is we would need to create an equivalent for other cloud providers as Azure WebJobs may not be available in other providers.

**Comment by bdukes** [comment]
It seems like the basic level would be @mitchelsellers' option 1, where the configuration of task specifies either :one: Run once on any server, :two: Run on every server, or :three: Run on a specific list of servers (what we have now, though maybe this should have the option to be a whitelist and blacklist).

However, if there's interest in environment-specific solutions, that would be a welcome addition to what we have now.  Ideally we have both.

There used to be a setting regarding whether the scheduler was timer based or request based, but I can't find it anymore.  I'm not sure if that was implemented in an extensible way, or was just two options supported by the main built-in scheduler.  However, there is a `scheduling` section in the `web.config` which, I would assume, could accept an Azure WebJobs implementation if one existed.  It might need an accompanying Connector to configure it.

**Comment by valadas** [comment]
Is this RFC discussed enough and ready to assign some milestone or we still need to leave it in RFC status ?

**Comment by mitchelsellers** [comment]
Assigning this to 10.x as I REALLY want it in there

**Comment by bdukes** [comment]
@mitchelsellers Am I right that this issue still doesn't have a solution?  If we're running a site on an Azure App Service, there is no way to force a job to only run on one server, since they don't have stable server names, correct?

**Comment by bdukes** [comment]
Does Evoq have a solution for this?  Or is this completely unsolved?

**Comment by mitchelsellers** [comment]
Yes, this is correct!  I've been chatting with @meetmandeep and his team is willing to contribute work to help with this.

Evoq does not properly solution for auto-scale either.

**Comment by mitchelsellers** [comment]
@meetmandeep If you have a minute, this is one of the ones I was mentioning a while back that we should chat about.
